### ISSUE TITLE
Add ability to create a temporary file with a specific extension to New-TemporaryFile

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#4215]
<sup>Related: #3442</sup>

Currently, `New-TemporaryFile` invariably creates temp. files with extension `.tmp`.

Sometimes, however, it is helpful to create temp. files with a specific filename extension.

Adding an `-Extension` parameter could provide that:

Desired behavior
------------------

```powershell
> New-TemporaryFile -Extension .ps1

    Directory: C:\Users\jdoe\AppData\Local\Temp

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        7/10/2017   9:53 PM              0 tmpE3B8.ps1

```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.3
```

**Comment by iSazonov** [comment]
What practical scenarios exist when do we need to change the extention?

**Comment by mklement0** [comment]
Whenever you need to create a temporary file to be processed by a utility that requires a specific filename extension.

Using PowerShell itself as an example, say you want to create a throw-away script dynamically, whose path you want to pass to another PowerShell instance via the CLI (`powershell -File ...`).  
Such a throw-away script only works if it has extension `*.ps1`.

**Comment by SteveL-MSFT** [comment]
@mklement0 with a recent PR, if you use -File, it doesn't have to have a .ps1 extension :)

**Comment by mklement0** [comment]
@SteveL-MSFT: Good point; I should have used a different example, such as `*.csv` for files to be processed by Excel, for instance.

**Comment by bergmeister** [comment]
I opened the pull request above for it. Please ignore the commits above except for the last one as I squashed the commits in my branch into one commit at the end.

**Comment by mklement0** [comment]
Thanks, @bergmeister!

**Comment by SimonWahlin** [comment]
PR #4612 was closed, waiting for functionality in CoreFX.
Will this cover both extension and total number of files cap?
Where can I vote for the API mentioned in [https://github.com/PowerShell/PowerShell/pull/4612#issuecomment-367381354](https://github.com/PowerShell/PowerShell/pull/4612#issuecomment-367381354)?

**Comment by bergmeister** [comment]
The corefx issue is here:
https://github.com/dotnet/corefx/issues/24001
The file cap limit is due to the limitation of the Windows API being used (and the short file name) for legacy reasons to avoid breaking changes. One would hope so that the new overload does not make use of that API any more, but please discuss this in the corefx repo/issue.

**Comment by markekraus** [comment]
@SteveL-MSFT Can I recommend a new tag that relates to awaiting upstream features and fixes in CoreFX? The closest we have is Waiting-Netstandard20 which isn't necessarily accurate for proposed but not-yet schedule changes such as the one this PR is pending.

**Comment by bergmeister** [comment]
At my workplace, we just mark it as blocked, sometimes that is easier instead of having multiple tags for why it is blocked unless there is a collection of issues being blocked on one thing like netstandard20

**Comment by iSazonov** [comment]
Yes, please rename.

**Comment by SteveL-MSFT** [comment]
Label has been renamed.

**Comment by AspenForester** [comment]
Given that I just had a situation where this functionality would have helped me, I would say that this is still relevant in in June 2020

**Comment by bergmeister** [comment]
Well it's currently on track for .Net 5

**Comment by Zian** [comment]
Since .Net 5 and PS 7.1 have already shipped while https://github.com/dotnet/runtime/issues/2048 is still "Open" with a milestone of "Future", does that mean this issue is effectively closed with the GitHub equivalent of WONTFIX?

**Comment by groutoutlook** [comment]
behold stale bot.

EDIT:
Please consider to support it. My use case is relatively simple, generate a temp file with generated content in stdin (html files) and open in with `msedge`. Edge can't take html stdin input so I reside to making a temp file, but it couldnt read `tmp` file there. So I need to create a proper file and delete them instead, which is unnecessary verbose.
